src/ctap/mod.rs

```diff
@@ -919,6 +923,90 @@ where
         Ok(ResponseData::AuthenticatorSelection)
     }
 
+    fn process_vendor_configure(
+        &mut self,
+        params: AuthenticatorVendorConfigureParameters,
+        cid: ChannelID,
+    ) -> Result<ResponseData, Ctap2StatusCode> {
+        (self.check_user_presence)(cid)?;
+
+        // Sanity checks
+        let current_priv_key = self.persistent_store.attestation_private_key()?;
+        let current_cert = self.persistent_store.attestation_certificate()?;
+
+        let response = if params.attestation_material.is_some() {
+            let data = params.attestation_material.unwrap();
+
+            match (current_cert, current_priv_key) {
+                (Some(_), Some(_)) => {
+                    // Device is fully programmed.
+                    // We don't compare values to avoid giving an oracle
+                    // about the private key.
+                    AuthenticatorVendorResponse {
+                        cert_programmed: true,
+                        pkey_programmed: true,
+                    }
+                }
+                // Device is not programmed.
+                (None, None) => {
+                    self.persistent_store
+                        .set_attestation_certificate(&data.certificate)?;
+                    self.persistent_store
+                        .set_attestation_private_key(&data.private_key)?;
+                    AuthenticatorVendorResponse {
+                        cert_programmed: true,
+                        pkey_programmed: true,
+                    }
+                }
+                // Device is partially programmed. Ensure the programmed value
+                // matched the given one before programming anything.
+                (Some(cert), None) => {
+                    if cert != data.certificate {
+                        return Err(Ctap2StatusCode::CTAP1_ERR_INVALID_PARAMETER);
+                    }
+                    self.persistent_store
+                        .set_attestation_private_key(&data.private_key)?;
+                    AuthenticatorVendorResponse {
+                        cert_programmed: true,
+                        pkey_programmed: true,
+                    }
+                }
+                (None, Some(key)) => {
+                    if key != data.private_key {
+                        return Err(Ctap2StatusCode::CTAP1_ERR_INVALID_PARAMETER);
+                    }
+                    self.persistent_store
+                        .set_attestation_certificate(&data.certificate)?;
+                    AuthenticatorVendorResponse {
+                        cert_programmed: true,
+                        pkey_programmed: true,
+                    }
+                }
+            }
+        } else {
+            AuthenticatorVendorResponse {
+                cert_programmed: current_cert.is_some(),
+                pkey_programmed: current_priv_key.is_some(),
+            }
+        };
```

Comment:
It looks like there is too much duplication. What about something like the following?

```rust
let response = match params.attestation_material {
    // Only reading values.
    None => Response { cert: cur_cert.is_some(), pkey: cur_pkey.is_some() },
    // Device is already fully programmed. We don't leak information.
    Some(_) if cur_cert.is_some() && cur_pkey.is_some() => Response { cert: true, pkey: true },
    // Device is partially or not programmed. We complete the process.
    Some(data) => {
        match cur_cert {
            None => store.set_cert(&data.cert)?,
            Some(cur_cert) if cur_cert == data.cert => (),
            Some(_) => return Err(INV_PARAM),
        }
        // Same for pkey
        Response { cert: true, pkey: true }
    }
};
```

Reply:
Wasn't aware that such complex conditions could be done :) 

One note though: I'm not sure both snippets are 100% equivalent: by treating the certificate first, we may program the cert in the store before checking that the priv key is matching.

Reply:
Oh good point. I guess in practice it doesn't matter since the pkey cannot be programmed if the cert is not. Alternatively for the last outer match branch we can do:

```rust
if let Some(cur_cert) = cur_cert {
    if cur_cert != data.cert {
        return Err(INV_PARAM);
    }
}
// Same for pkey
if cur_cert.is_none() {
    store.set_cert(&data.cert)?;
}
// Same for pkey
```

Reply:
Done.